File: leetcode/1002.find-common-characters/cpp/commonchars.cpp
#include <algorithm>
#include <array>
#include <string>
#include <vector>

class Solution
{
public:
    std::vector<std::string> commonChars(
        const std::vector<std::string> &words)
    {
        if (words.empty())
            return {};

        auto c = countWord(words[0]);

        for (std::size_t i = 1; i < words.size(); ++i)
        {
            const auto cl = countWord(words[i]);

            for (int j = 0; j < 26; ++j)
                c[j] = std::min(c[j], cl[j]);
        }

        std::vector<std::string> res;

        for (int i = 0; i < 26; ++i)
        {
            for (int j = 0; j < c[i]; ++j)
                res.push_back(std::string(1, static_cast<char>('a' + i)));
        }

        return res;
    }

private:
    std::array<int, 26> countWord(const std::string &word)
    {
        std::array<int, 26> counts{};

        for (const char ch : word)
            counts[ch - 'a']++;

        return counts;
    }
};

File: leetcode/1002.find-common-characters/ts/commonchars.ts
const commonChars = (words: string[]): string[] => {
    if (words.length === 0) return []

    const c = countWord(words[0])

    for (const w of words.slice(1)) {
        const cl = countWord(w)

        for (let i = 0; i < 26; i++) {
            c[i] = Math.min(c[i], cl[i])
        }
    }

    const res: string[] = []

    for (let i = 0; i < 26; i++) {
        for (let j = 0; j < c[i]; j++) {
            res.push(String.fromCharCode('a'.charCodeAt(0) + i))
        }
    }

    return res
}


const countWord = (word: string): number[] => {
    const c: number[] = Array(26).fill(0)

    for (const ch of word) {
        const i = ch.charCodeAt(0) - 'a'.charCodeAt(0)
        c[i]++
    }

    return c
}

export default commonChars
